main.go

```diff
@@ -1,152 +1,105 @@
 package main
 
 import (
-	"context"
-	"encoding/json"
 	"fmt"
-	"io"
-	"io/ioutil"
 	"net"
 	"os"
 	"os/signal"
 	"syscall"
 
-	log "github.com/sirupsen/logrus"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/secrets-store-csi-driver-provider-vault/internal/providerserver"
+	"github.com/hashicorp/secrets-store-csi-driver-provider-vault/internal/version"
 	"github.com/spf13/pflag"
 	"google.golang.org/grpc"
 	pb "sigs.k8s.io/secrets-store-csi-driver/provider/v1alpha1"
 )
 
 var (
-	endpoint = pflag.String("endpoint", "/tmp/vault.sock", "path to socket on which to listen for driver gRPC calls")
-	debug    = pflag.Bool("debug", false, "sets log to debug level")
-	version  = pflag.Bool("version", false, "prints the version information")
+	endpoint    = pflag.String("endpoint", "/tmp/vault.sock", "path to socket on which to listen for driver gRPC calls")
+	debug       = pflag.Bool("debug", false, "sets log to debug level")
+	selfVersion = pflag.Bool("version", false, "prints the version information")
 )
 
-const minDriverVersion = "v0.0.17"
-
-// LogHook is used to setup custom hooks
-type LogHook struct {
-	Writer    io.Writer
-	Loglevels []log.Level
+func main() {
+	logger := hclog.Default()
+	err := realMain(logger)
+	if err != nil {
+		logger.Error("Error running provider", "err", err)
+		os.Exit(1)
+	}
 }
 
-func main() {
+func realMain(logger hclog.Logger) error {
 	pflag.Parse()
 
-	setupLogger()
+	// set log level
+	logger.SetLevel(hclog.Info)
+	if *debug {
+		logger.SetLevel(hclog.Debug)
+	}
 
-	if *version {
-		v, err := getVersion()
+	if *selfVersion {
+		v, err := version.GetVersion()
 		if err != nil {
-			log.Fatalf("failed to print version, err: %+v", err)
+			return fmt.Errorf("failed to print version, err: %w", err)
 		}
 		// print the version and exit
-		fmt.Printf("%s\n", v)
-		os.Exit(0)
+		logger.Info(v)
+		return nil
 	}
 
+	logger.Info("Creating new gRPC server")
 	server := grpc.NewServer()
 
 	c := make(chan os.Signal, 1)
 	signal.Notify(c, syscall.SIGTERM, syscall.SIGINT)
 	go func() {
 		sig := <-c
-		log.Infof("Caught signal %s, shutting down", sig)
+		logger.Info(fmt.Sprintf("Caught signal %s, shutting down", sig))
 		server.GracefulStop()
 	}()
 
-	listener, err := net.Listen("unix", *endpoint)
+	listener, err := listen(logger)
 	if err != nil {
-		log.Fatalf("Failed to listen on unix socket at %s: %v", *endpoint, err)
+		return err
 	}
 	defer listener.Close()
-	log.Infof("Listening on %s", *endpoint)
+	logger.Info(fmt.Sprintf("Listening on %s", *endpoint))
 
-	s := &ProviderServer{}
+	s := &providerserver.ProviderServer{
+		Logger: logger,
+	}
 	pb.RegisterCSIDriverProviderServer(server, s)
 
 	err = server.Serve(listener)
 	if err != nil {
-		log.Fatal(err)
+		return fmt.Errorf("error running gRPC server: %w", err)
```

Comment:
Should we also warn using the logger here?

Reply:
`main()` will log this as an error-level log before the provider exits with exit code 1.